lib/index.ts

```diff
@@ -57,6 +57,141 @@ export async function inspect(
 }
 
+interface UpgradeRemediation {
+  upgradeTo: string;
+  // Other fields are of no interest
+}
+
+interface DependencyUpdates {
+  [from: string]: UpgradeRemediation;
+}
+
+interface ManifestFiles {
+  // Typically these are requirements.txt, constraints.txt and Pipfile;
+  // the plugin supports paths with subdirectories
+  [name: string]: string; // name-to-content
+}
+
+// Correction for the type; should be fixed in snyk-cli-interface
+export interface DepTreeDep {
+  name?: string; // shouldn't, but might be missing
+  version?: string; // shouldn't, but might be missing
+  dependencies?: {
+    [depName: string]: DepTreeDep;
+  };
+  labels?: {
+    [key: string]: string;
+
+    // Known keys:
+    // pruned: identical subtree already presents in the parent node.
+    //         See --prune-repeated-subdependencies flag.
+  };
+}
+
+// Applies upgrades to direct and indirect dependencies
+export async function applyRemediationToManifests(
+  root: string,
+  manifests: ManifestFiles,
+  upgrades: DependencyUpdates,
+  options: Options
+) {
+  const manifestNames = Object.keys(manifests);
+  const targetFile = manifestNames.find(
+    (fn) => path.basename(fn) === 'requirements.txt'
+  );
+  if (
+    !targetFile ||
+    !manifestNames.every(
+      (fn) =>
+        path.basename(fn) === 'requirements.txt' ||
+        path.basename(fn) === 'constraints.txt'
+    )
+  ) {
+    throw new Error(
+      'Remediation only supported for requirements.txt and constraints.txt files'
+    );
+  }
+
+  const provOptions = { ...options };
+  provOptions.args = provOptions.args || [];
+  provOptions.args.push('--only-provenance');
+
+  const topLevelDeps = (await inspect(root, targetFile, provOptions)).package;
+  applyUpgrades(manifests, upgrades, topLevelDeps);
+
+  return manifests;
+}
+
+function applyUpgrades(
+  manifests: ManifestFiles,
+  upgrades: DependencyUpdates,
+  topLevelDeps: DepTree
+) {
+  const requirementsFileName = Object.keys(manifests).find(
+    (fn) => path.basename(fn) === 'requirements.txt'
+  ) as string;
+  const constraintsFileName = Object.keys(manifests).find(
+    (fn) => path.basename(fn) === 'constraints.txt'
+  );
+
+  // Updates to requirements.txt
+  const patch: { [zeroBasedIndex: number]: string | false } = {}; // false means remove the line
+  const append: string[] = [];
+
+  for (const upgradeFrom of Object.keys(upgrades)) {
+    const pkgName = upgradeFrom.split('@')[0].toLowerCase();
+    const newVersion = upgrades[upgradeFrom].upgradeTo.split('@')[1];
+    const topLevelDep = (topLevelDeps.dependencies || {})[
+      pkgName
+    ] as DepTreeDep;
+    if (topLevelDep && topLevelDep.labels && topLevelDep.labels.provenance) {
+      const lineNumbers = topLevelDep.labels.provenance
+        .split(':')[1]
+        .split('-')
+        .map((x) => parseInt(x));
+      // TODO: what if the original version spec was range, e.g. >=1.0,<2.0 ?
+      patch[lineNumbers[0] - 1] = pkgName + '==' + newVersion;
```

Comment:
what does patch mean here? Because in remediation it means a fix patch, perhaps a different word is better since I assume we don't have patches yet for Python?

Reply:
There's a comment right next to the definition of `patch`.